### PR TITLE
Reviewer Rob - Simple addition of the API

### PR DESCRIPTION
**Pull request by bossmc**
Not yet UTd/regression tested (that's been pushed to next week), but can you please review this change to add the API we previously discussed to the MemcachedStoreView?

### CODE REVIEW

**Comment by bossmc** [comment]
@AlexHockey FYI

File: include/memcachedstoreview.h [hunk 1]
```diff
@@ -41,6 +41,7 @@
 #include <map>
 #include <string>
 #include <vector>
+#include <list>
```
Comment (rkd-msw) [inline include/memcachedstoreview.h]: Why?

File: include/memcachedstoreview.h [hunk 2]
```diff
@@ -61,6 +62,20 @@ class MemcachedStoreView
...
   const std::vector<std::string>& write_replicas(int vbucket) const { return _write_set[vbucket]; };
 
+  /// Calculates the vbucket moves that are currently ongoing.
+  ///
+  /// The returned object has an entry for each moving vbucket ID, giving the
+  /// old the replica list and the new one.  vBuckets that are not moving are
+  /// skipped in the output (thus, if there's no move ongoing, this map is
+  /// empty).
+  const std::map<int,
+                 std::pair<std::vector<std::string>,
```
Comment (rkd-msw) [inline include/memcachedstoreview.h]: Can this be typedef'ed to be clearer?

File: include/memcachedstoreview.h [hunk 3]
```diff
@@ -61,6 +62,20 @@ class MemcachedStoreView
   const std::vector<std::string>& read_replicas(int vbucket) const { return _read_set[vbucket]; };
   const std::vector<std::string>& write_replicas(int vbucket) const { return _write_set[vbucket]; };
 
+  /// Calculates the vbucket moves that are currently ongoing.
+  ///
+  /// The returned object has an entry for each moving vbucket ID, giving the
+  /// old the replica list and the new one.  vBuckets that are not moving are
```
Comment (rkd-msw) [inline include/memcachedstoreview.h]: "the old the replica list"

File: src/memcachedstoreview.cpp [hunk 4]
```diff
@@ -132,6 +139,33 @@ void MemcachedStoreView::update(const std::vector<std::string>& servers,
       std::vector<int> c_nodes = c_ring.get_nodes(ii, _replicas);
       std::vector<int> n_nodes = n_ring.get_nodes(ii, _replicas);
 
+      // Determine if the set of nodes has changed by sorting the above two
+      // vectors and comparing.
+      std::vector<int> c_nodes_sorted = c_nodes;
```
Comment (rkd-msw) [inline src/memcachedstoreview.cpp]: What's the c and n prefix? Is it 'current' and 'new'?

File: src/memcachedstoreview.cpp [hunk 5]
```diff
@@ -132,6 +139,33 @@ void MemcachedStoreView::update(const std::vector<std::string>& servers,
...
+      std::vector<int> n_nodes_sorted = n_nodes;
+      std::sort(c_nodes_sorted.begin(), c_nodes_sorted.end());
+      std::sort(n_nodes_sorted.begin(), n_nodes_sorted.end());
+      if (c_nodes_sorted != n_nodes_sorted)
+      {
+        // Lists are different, add an entry to _changes to indicate this.
+        std::vector<std::string> c_replicas;
+        std::vector<std::string> n_replicas;
+        for (std::vector<int>::const_iterator it = c_nodes.begin();
```
Comment (rkd-msw) [inline src/memcachedstoreview.cpp]: Maybe worth a comment to explain what's going on here (index -> server name string).

**Comment by rkd-msw** [comment]
Reviewed.

File: include/memcachedstoreview.h [hunk 6]
```diff
@@ -129,6 +142,9 @@ class MemcachedStoreView
   // replicas are enabled to maintain redundancy.
   std::vector<std::vector<std::string> > _read_set;
   std::vector<std::vector<std::string> > _write_set;
+
+  // vBucket allocation hanges currently ongoing in the cluster (may be empty).
```
Comment (AlexHockey) [inline include/memcachedstoreview.h]: Typo "hanges"

File: include/memcachedstoreview.h [hunk 7]
```diff
@@ -129,6 +142,9 @@ class MemcachedStoreView
...
+  std::map<int, std::pair<std::vector<std::string>, std::vector<std::string>>> _changes;
```
Comment (AlexHockey) [inline include/memcachedstoreview.h]: Why not use the typedefs defined above?

**Pull request merged**